### PR TITLE
Further API-capabilities for plugins

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## OpenProject Work Package

https://community.openproject.org/work_packages/19044
## Summary

This PR adds further capabilities required by https://github.com/finnlabs/openproject-backlogs/pull/123.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-even-more-capabilities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-even-more-capabilities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: lib/api/decorators/schema.rb [hunk 1]
```diff
@@ -48,8 +48,8 @@ def schema(property,
                      name = call_or_translate(name_source)
                      schema = ::API::Decorators::PropertySchemaRepresenter.new(type: type,
                                                                                name: name,
-                                                                               required: required,
-                                                                               writable: writable)
+                                                                               required: call_or_use(required),
```
Comment (NobodysNightmare) [inline lib/api/decorators/schema.rb]: I hate that this line got this long, but it is basically only this darn namespace prefix...

I'd rather keep it as long if there are no objections...

Comment (NobodysNightmare) [inline lib/api/decorators/schema.rb]: @Azure7111 told me how to do best... I agree ^^

File: lib/api/v3/work_packages/work_package_contract.rb [hunk 2]
```diff
@@ -27,28 +27,39 @@
...
-            due_date
-            status_id
-            assigned_to_id
-            responsible_id
-            priority_id
-            category_id
-            fixed_version_id
-          )
+      class WorkPackageContract < ::API::Contracts::ModelContract
+        writable_attribute :subject
```
Comment (ulferts) [inline lib/api/v3/work_packages/work_package_contract.rb]: just a naming suggesting: we only have `writable_attributes` - a whitelist. Every attribute not listed is not writeable. Can't we just shorten this to `attribute`. That would make it easier for me to understand the validation block as I would no longer think that the block has something to do with the writing part.

Comment (ulferts) [inline lib/api/v3/work_packages/work_package_contract.rb]: Additionally, we could allow for specifying multiple attributes at the same time.

File: lib/open_project/plugins/acts_as_op_engine.rb [hunk 3]
```diff
@@ -68,6 +68,18 @@ def self.included(base)
...
+        plugin_name = engine_name
+        base.config.to_prepare do
+          klass_name = args.last
+          plugin_module = plugin_name.sub(/^openproject_/, '').camelcase
+          patch = "OpenProject::#{plugin_module}::Patches::#{klass_name}Patch".constantize
+          qualified_class_name = args.map(&:to_s).join('::')
+          klass = qualified_class_name.to_s.constantize
+          klass.send(:include, patch) unless klass.included_modules.include?(patch)
+        end
+      end
```
Comment (ulferts) [inline lib/open_project/plugins/acts_as_op_engine.rb:79]: You can get the patches constant easier and more robust like this:

```
plugin_module = self.class.to_s.deconstantize
base.config.to_prepare do
  klass_name = args.last
  patch = "#{plugin_module}::Patches::#{klass_name}Patch".constantize
  ....
end
```

And while you are at it: **you should not copy code**

**Comment by ulferts** [comment]
Functionality looks good, should be possible to merge this, soon.

File: lib/api/v3/work_packages/work_package_contract.rb (hunk 2 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/work_package_contract.rb]: Sounds good to me,

however I'd rather not use that feature too extensively, because I feel that the readability does not increase very much, when you put a bunch of unrelated attributes into one line.

File: lib/open_project/plugins/acts_as_op_engine.rb (hunk 3 above)
Comment (NobodysNightmare) [inline lib/open_project/plugins/acts_as_op_engine.rb:79]: I thought this is how programming works?!?

copy & paste

**Comment by ulferts** [comment]
@NobodysNightmare rebasing required. 
image

**Comment by NobodysNightmare** [comment]
Travis says it is green:

green as the sun

**Comment by ulferts** [comment]
@NobodysNightmare <- Windows user.
